circup/__init__.py

```diff
@@ -409,6 +410,616 @@ def __repr__(self):
+                        source or from a pre-compiled module
+        :param mod_names: Dictionary of metadata from modules that can be generated
+                           with get_bundle_versions()
+        """
+        if not name:
+            click.echo("No module name(s) provided.")
+        elif name in mod_names:
+            # Grab device modules to check if module already installed
+            if name in device_modules:
+                click.echo("'{}' is already installed.".format(name))
+                return
+
+            library_path = os.path.join(device_path, self.LIB_DIR_PATH)
+
+            # Create the library directory first.
+            self._create_library_directory(device_path, library_path)
+
+            metadata = mod_names[name]
+            bundle = metadata["bundle"]
+            if pyext:
+                # Use Python source for module.
+                self._install_module_py(library_path, metadata)
+            else:
+                # Use pre-compiled mpy modules.
+                self._install_module_mpy(bundle, library_path, metadata)
+            click.echo("Installed '{}'.".format(name))
+        else:
+            click.echo("Unknown module named, '{}'.".format(name))
+
+    def libraries_from_imports(self, code_py, mod_names):
+        """
+        To be overridden by subclass
+        """
+        raise NotImplementedError
+
+    def uninstall(self, device_path, module_path):
+        """
+        To be overridden by subclass
+        """
+        raise NotImplementedError
+
+    def update(self, module):
+        """
+        To be overridden by subclass
+        """
+        raise NotImplementedError
+
+    def libraries_from_code_py(self, code_py, mod_names):
+        """
+        Parse the given code.py file and return the imported libraries
+
+        :param str code_py: Full path of the code.py file
+        :return: sequence of library names
+        """
+        # pylint: disable=broad-except
+        try:
+            found_imports = findimports.find_imports(code_py)
+        except Exception as ex:  # broad exception because anything could go wrong
+            logger.exception(ex)
+            click.secho('Unable to read the auto file: "{}"'.format(str(ex)), fg="red")
+            sys.exit(2)
+        # pylint: enable=broad-except
+        imports = [info.name.split(".", 1)[0] for info in found_imports]
+        return [r for r in imports if r in mod_names]
+
+
+class WebBackend(Backend):
+    """
+    Backend for interacting with a device via Web Workflow
+    """
+
+    def __init__(self, host, password):
+        super().__init__()
+        self.LIB_DIR_PATH = "fs/lib/"
+        self.host = host
+        self.password = password
+
+    def install_file_http(self, source, target):
+        """
+        Install file to device using web workflow.
+        :param source source file.
+        :param target destination URL. Should have password embedded.
```

Comment:
Having the full url and password in the target URL seems like a leaky abstraction. Instead, the paths to each backend should be the same.

Reply:
latest commit removes `target` argument entirely and derives the URL from some values from `self` & `source`

Same change made in `install_dir_http()` as well.